pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,241 @@
+        self.m_best_val =- 1000000
+        # This is to be used for calculating the breaking point of RunMPLP
+        self.last_obj = 1e40
+        for i in range(reader.no_of_functions):
+            self.m_sum_into_intersects.append([log(y) for y in all_lambdas[i]])
+            if len(all_factors[i])>1:
+                self.m_region_lambdas.append(self.m_sum_into_intersects[i])
+                curr_region = Region(all_factors[i], all_factors[:i+1], all_factors[i], all_cardinalities, i)
+                self.m_all_regions.append(curr_region)
+
+            # Initialise output vector
+            self.m_decoded_res = [0]*len(all_cardinalities)
+            self.m_best_decoded_res = [0]*len(all_cardinalities)
+
+    # To convert a [1 0 0] to its flat value 4 if the cardinality for each variable is [2 2 2]
+    # Similarly to convert a [1 0 0] to 6 if the cardinality foreach variable is [4 3 2]
+    def GetFlatInd(self, base_ind_for_multi_index, region_base_sizes):
+        return [list(i) for i in itertools.product(*[list(range(k)) for k in region_base_sizes])].index(base_ind_for_multi_index)
+
+    # To see if m_decoded _res indeed is the m_best_decoded_res, we need to calculate int_val.
+    def IntVal(self):
+        int_val=0
+        for i in range(len(self.m_all_regions)):
+            tmpvec = []
+            base_sizes = []
+            for j in self.m_all_regions[i].m_region_inds:
+                tmpvec.append(self.m_decoded_res[j])
+                base_sizes.append(self.m_var_sizes[j])
+            int_val += self.m_region_lambdas[i][self.GetFlatInd(tmpvec, base_sizes)]
+        return int_val
+
+    # Find an integral solution x by locally maximizing the single node beliefs.
+    # Here x = m_best_decoded_res
+    def LocalDecode(self):
+        obj=0
+        for i in range(len(self.m_sum_into_intersects)):
+            obj+=max(self.m_sum_into_intersects[i])
+            index_of_maximum=self.m_sum_into_intersects[i].index(max(self.m_sum_into_intersects[i]))
+            if(len(self.m_all_intersect[i])==1):
+                self.m_decoded_res[self.m_all_intersect[i][0]]=index_of_maximum
+
+        # update the m_best_val if it is smaller than int_val
+        int_val=self.IntVal()
+        print('int_val: "{}"'.format(int_val))
+
+        if(int_val>self.m_best_val):
+            self.m_best_decoded_res=self.m_decoded_res
+            self.m_best_val=int_val
+        return obj
+
+    def RunMPLP(self, niter, obj_del_thr, int_gap_thr):
+        # Perform the GMPLP updates.
+        for i in range(niter):
+            # For each of the 24 regions
+            for r in self.m_all_regions:
+                self.m_sum_into_intersects=r.UpdateMsgs(self.m_sum_into_intersects)
+            obj=self.LocalDecode()
+            obj_del= self.last_obj-obj
+            self.last_obj=obj
+            int_gap=obj-self.m_best_val
+
+            if (obj_del<obj_del_thr) and (i>16) :
```

Comment:
What does ```16``` signify. Never put the constants directly. Put these constants in a variable ```VAR_NAME``` and use instead of it. It is more robust and easier to read the code.